activerecord/lib/active_record/connection_adapters/mysql_adapter.rb

```diff
@@ -407,24 +345,18 @@ def exec_stmt(sql, name, binds, cache_stmt: false)
             if !cache_stmt
               stmt.close
             else
-              @statements.delete sql
+              @statements.delete(sql)
             end
             raise e
           end
 
-          cols = nil
-          if metadata = stmt.result_metadata
-            cols = cache[:cols] ||= metadata.fetch_fields.map(&:name)
-            metadata.free
-          end
-
-          result_set = ActiveRecord::Result.new(cols, stmt.to_a) if cols
-          affected_rows = stmt.affected_rows
+          result = stmt.result_metadata
+          ret = yield stmt, result
+          result.free if result
 
           stmt.free_result
```

Comment:
Oh, I didn't see your reply before I edited my own comment :) I'd forgotten that the result set and the result metadata are separate items.